spec.emu

```diff
@@ -882,6 +882,297 @@ contributors: Google, Ecma International
+          1. Set _daysInMonth_ to CalendarDaysInMonth(_calendar_, _resolvedYear_, _resolvedMonth_).
+          1. Set _resolvedDay_ to _day_ + _daysInMonth_.
+        1. Repeat, while _resolvedDay_ &gt; _daysInMonth_,
+          1. Set _resolvedDay_ to _resolvedDay_ - _daysInMonth_.
+          1. Set _resolvedMonth_ to _resolvedMonth_ + 1.
+          1. If _resolvedMonth_ &gt; _monthsInYear_, then
+            1. Set _resolvedYear_ to _resolvedYear_ + 1.
+            1. Set _monthsInYear_ to CalendarMonthsInYear(_calendar_, _resolvedYear_).
+            1. Set _resolvedMonth_ to 1.
+          1. Set _daysInMonth_ to CalendarDaysInMonth(_calendar_, _resolvedYear_, _resolvedMonth_).
+        1. Return the Record { [[Year]]: _resolvedYear_, [[Month]]: _resolvedMonth_, [[Day]]: _resolvedDay_ }.
+      </emu-alg>
+    </emu-clause>
+
+    <emu-clause id="sec-temporal-nonisodatesurpasses" type="implementation-defined abstract operation">
+      <h1>
+        NonISODateSurpasses (
+          _calendar_: a calendar type that is not *"iso8601"*,
+          _sign_: -1 or 1,
+          _baseDate_: an ISO Date Record,
+          _isoDate2_: an ISO Date Record,
+          _years_: an integer,
+          _months_: an integer,
+          _weeks_: an integer,
+          _days_: an integer,
+        ): a Boolean
+      </h1>
+      <dl class="header">
+        <dt>description</dt>
+        <dd>
+          The return value indicates whether the date _date1_, the result of adding the duration denoted by _years_, _months_, _weeks_, and _days_ to _baseDate_ in the calendar system denoted by _calendar_, surpasses _isoDate2_ in the direction denoted by _sign_.
+          If _weeks_ and _days_ are both zero, then _date1_ need not exist (for example, it could be February 30).
+        </dd>
+      </dl>
+      <p>It performs the following steps when called:</p>
+      <emu-alg>
+        1. Let _parts_ be CalendarISOToDate(_calendar_, _baseDate_).
+        1. Let _yearMonth_ be RegulateNonISOYearMonth(_calendar_, _parts_.[[Year]], _parts_.[[MonthCode]], _years_).
+        1. Let _endOfMonth_ be BalanceNonISODate(_calendar_, _yearMonth_.[[Year]], _yearMonth_.[[Month]] + _months_ + 1, 0).
+        1. Let _baseDay_ be _parts_.[[Day]].
+        1. If _weeks_ is not 0 or _days_ is not 0, then
+          1. If _baseDay_ &lt; _endOfMonth_.[[Day]], then
+            1. Let _regulatedDay_ be _baseDay_.
+          1. Else,
+            1. Let _regulatedDay_ be _endOfMonth_.[[Day]].
+          1. Let _balancedDate_ be BalanceNonISODate(_calendar_, _endOfMonth_.[[Year]], _endOfMonth_.[[Month]], _regulatedDay_ + 7 * _weeks_ + _days_).
+          1. Let _y1_ be _balancedDate_.[[Year]].
+          1. Let _m1_ be _balancedDate_.[[Month]].
+          1. Let _d1_ be _balancedDate_.[[Day]].
+        1. Else,
+          1. Let _y1_ be _endOfMonth_.[[Year]].
+          1. Let _m1_ be _endOfMonth_.[[Month]].
+          1. Let _d1_ be _baseDay_.
```

Comment:
issue: this doesn't clamp, but the previous one does. Is that a discrepancy we are okay with?

This means `Surpasses(calendar, sign = 1, baseDate, isoDate2, years, months, weeks = 0, days = 0)` could be true whilst `Surpasses(calendar, sign = 1, baseDate, isoDate2, years, months, weels = 0, days = 1)` could be false if, say, the other numbers were something like "add 1 month to to January 31st" since with days = 1 it will get normalized to Feb 28 + 1 (assume non leap year) but with days = 0 it'll get normalized to Feb 32.

Reply:
This matches the behavior in https://github.com/tc39/proposal-temporal/pull/3138.

The AO has one call site, `Until`, which wants a `true` return value for things like "march 31 + 1 month ?< april 30", since the Temporal Addition Algorithm only regulates to the end of the month immediately before adding weeks and days.

I could have written this AO with an extra argument, `regulationBehavior` or something, or even as a completely different AO, but I realized that checking `weeks == 0 and days == 0` was equivalent.